### ISSUE TITLE
[Discuss] Move StoreBlob properties to SupportedPaymentmethod

### ISSUE THREAD

**Issue by Kukks** [btcpayserver/btcpayserver#1621]
After seeing discussion on https://github.com/btcpayserver/btcpayserver/pull/1620, I've come to realize that we should try and break up the big blobs that hold many properties into sub resources. This way, we are less prone to require version bumps 

Most of these properties do not even belong on the store level and should be on the payment method anyway.
Here is the current greenfield mapping for a store
```
                Id = data.Id, 
                Name = data.StoreName,
                Website = data.StoreWebsite,
                SpeedPolicy = data.SpeedPolicy,
                //we do not include the default payment method in this model and instead opt to set it in the stores/storeid/payment-methods endpoints
                //blob
                //we do not include DefaultCurrencyPairs,Spread, PreferredExchange, RateScripting, RateScript  in this model and instead opt to set it in stores/storeid/rates endpoints
                //we do not include ChangellySettings in this model and instead opt to set it in stores/storeid/changelly endpoints
                //we do not include CoinSwitchSettings in this model and instead opt to set it in stores/storeid/coinswitch endpoints
                //we do not include ExcludedPaymentMethods in this model and instead opt to set it in stores/storeid/payment-methods endpoints
                //we do not include EmailSettings in this model and instead opt to set it in stores/storeid/email endpoints
                //we do not include OnChainMinValue and LightningMaxValue because moving the CurrencyValueJsonConverter to the Client csproj is hard and requires a refactor (#1571 & #1572)
                NetworkFeeMode = storeBlob.NetworkFeeMode,
                RequiresRefundEmail = storeBlob.RequiresRefundEmail,
                ShowRecommendedFee = storeBlob.ShowRecommendedFee,
                RecommendedFeeBlockTarget = storeBlob.RecommendedFeeBlockTarget,
                DefaultLang = storeBlob.DefaultLang,
                MonitoringExpiration = storeBlob.MonitoringExpiration,
                InvoiceExpiration = storeBlob.InvoiceExpiration,
                LightningAmountInSatoshi = storeBlob.LightningAmountInSatoshi,
                CustomLogo = storeBlob.CustomLogo,
                CustomCSS = storeBlob.CustomCSS,
                HtmlTitle = storeBlob.HtmlTitle,
                AnyoneCanInvoice = storeBlob.AnyoneCanInvoice,
                LightningDescriptionTemplate = storeBlob.LightningDescriptionTemplate,
                PaymentTolerance = storeBlob.PaymentTolerance,
                RedirectAutomatically = storeBlob.RedirectAutomatically,
                PayJoinEnabled = storeBlob.PayJoinEnabled,
                LightningPrivateRouteHints = storeBlob.LightningPrivateRouteHints
```

I propose that some of these are moved to sub properties:
* `SpeedPolicy` - to `DerivationSchemeSettings` and `MoneroSupportedPaymentMethod`.  

If you accept alts, you would most likely want to have a speed policy to be higher for them. This field also makes no sense for `LightningSupportedPaymentMethod`

* `NetworkFeeMode`, `ShowRecommendedFee`, `RecommendedFeeBlockTarget`, `MonitoringExpiration`, `PayJoinEnabled`  - again only applies to  `DerivationSchemeSettings` and `MoneroSupportedPaymentMethod`
* `LightningAmountInSatoshi`, `LightningDescriptionTemplate`, `LightningPrivateRouteHints` - only applies to `LightningSupportedPaymentMethod`
* `DefaultCurrencyPairs`,`Spread`, `PreferredExchange`, `RateScripting`, RateScript` - belong in a `RatesSettings` sub object

**Comment by NicolasDorier** [comment]
There is no need of version bump for adding properties.

**Comment by NicolasDorier** [comment]
> SpeedPolicy - to DerivationSchemeSettings and MoneroSupportedPaymentMethod.

No, DerivationSchemeSettings is a wallet concept, speedpolicy is a store concept.

The API is fine as it is right now, there is no need to move things around.

**Comment by Kukks** [comment]
> 
> 
> There is no need of version bump for adding properties.

If you add a property, then the client needs to adapt to send that property or validation fails. Doesn't that warrant a version bump?

**Comment by Kukks** [comment]
> 
> 
> > SpeedPolicy - to DerivationSchemeSettings and MoneroSupportedPaymentMethod.
> 
> No, DerivationSchemeSettings is a wallet concept, speedpolicy is a store concept.
> 
> The API is fine as it is right now, there is no need to move things around.

Yes, and in #1570 I tried to treat it as such. I actually believe `ISupportedPaymentMethod` should be moved to its own dedicated table (like in that draft PR) and then have the `StoreData` have a list of ids linking to them.

So instead of having `public string DerivationStrategies { get; set; }`  ins StoreData, we would have:

```
public List<string> WalletIds{ get; set; }
public Dictionary<string, IWalletSettings> StoreWalletSettings { get; set; }
```

where `IWalletSettings` is something like
```
class LightningStoreWalletSettings : IWalletSettings{
        public bool LightningPrivateRouteHints { get; set; }
        public CurrencyValue LightningMaxValue { get; set; }
        public bool LightningAmountInSatoshi { get; set; }
}

class DerivationStoreWalletSettings : IWalletSettings{
        public bool ShowRecommendedFee { get; set; }
        public int RecommendedFeeBlockTarget { get; set; }
        public bool PayJoinEnabled{ get; set; }
        [JsonConverter(typeof(StringEnumConverter))]
        public NetworkFeeMode NetworkFeeMode { get; set; }
}

....
```


Why?
This lets us not just decouple wallet from store but also lets us have a wallet abstraction layer that lets us create wallets for all payment methods and types (and more)

**Comment by NicolasDorier** [comment]
Before talking about deep refactoring, I prefer we focus on finishing what is started and finish the current milestone.

* Default values should be reviewed and probably fixed on payment request API
* Documentation of all the fields in payment request API

If you think some field will soon be refactored and moved, maybe we should remove them before the next release so we don't have to maintain backward compatibility. 

But I think this refactoring is not simple and will take several months, as even the UI need to be redone, so I would prefer we just keep how things are now.

**Comment by NicolasDorier** [comment]
I thought about making a new version in the route as @rockstardev I don't think that's that bad actually. Just mean we have an additional parameter in the route that we need to check whether to ignore some fields or not.

**Comment by rockstardev** [comment]
I've sent you response on that other issue, building up on your feedback. The more I read, the more I think you are looking for client versioning which would allow server to process requests in different manner while maintaining compatibility.

Route parameters are tricky... usually people end up with having:
 - `SomethingSomething`
 - `SomethingSomethingV2`

Whenever I see that I know whoever designed the API could've benefited from more rockstarism in the team. I mean, it's quick fix that works... but strong client versioning is way more robust. Devs just don't want to deal with overhead, even when you explain the benefits - they prefer to keep everything on the server and that new route makes is more obvious in documentation.

So, I'll see what you two decide. But please no `AdditionalData`.

**Comment by NicolasDorier** [comment]
@rockstardev the only thing I am trying to prevent with #1621 is that a client project does not end up inadvertently overriding new fields after he updated his BTCPayServer.

Basically without my change, updating the server without the client, would make the client start breaking things, which should not happen.

Bumping version for each route for each change is possible, though overkill in my opinion. Also I would want to prevent situation where people making a client library get burned because they added the new field but forgot to bump the version, and thus the field get ignored.

**Comment by Kukks** [comment]
> 
> 
> @rockstardev the only thing I am trying to prevent with #1621 is that a client project does not end up inadvertently overriding new fields after he updated his BTCPayServer.
> 
> Basically without my change, updating the server without the client, would make the client start breaking things, which should not happen.

Unless you bump the api versioning :) I think `AdditionalData` on the client side is harmless, It allows client users to not wait for the library be updated but I do not think it should be used as a way to avoid version bumps.
 
> Bumping version for each route for each change is possible, though overkill in my opinion. Also I would want to prevent situation where people making a client library get burned because they added the new field but forgot to bump the version, and thus the field get ignored.

If you don't want to bump the api each time, then we need to apply a variation of
```
[JsonProperty(Required = Required.AllowNull)]
```
for each new property on the endpoint so that clients using the same version do not break.
```
    public enum Required
    {
        /// <summary>
        /// The property is not required. The default state.
        /// </summary>
        Default = 0,

        /// <summary>
        /// The property must be defined in JSON but can be a null value.
        /// </summary>
        AllowNull = 1,

        /// <summary>
        /// The property must be defined in JSON and cannot be a null value.
        /// </summary>
        Always = 2,

        /// <summary>
        /// The property is not required but it cannot be a null value.
        /// </summary>
        DisallowNull = 3
    }
}
```

The reason i created this issue is to reduce the need for bumping apis in case of property changes and to isolate properties to sub resources on the API (make models smaller). This way, breaking changes can have a smaller impact.

> On unrelated note, I will, I think, soon start removing all the `[Obsolete]` properties around in the project and instead make them read directly from an AdditionalData field. So the legacy does not pollute intellisense.

Awesome, ACK from my end

**Comment by rockstardev** [comment]
The problem with `AdditionalData` is that you are now pushing this non-intuitive pattern on all clients. I'm completely lost on why this is needed.

My responses included numerous options - some of them were about versioning on client, some versioning on client+server, some versioning on server, some without ANY versioning. So just to be clear on that... since I don't enjoy my suggestions being miss-represented and trivialized.

**Comment by rockstardev** [comment]
Agree... and I would go further by saying that I don't think `AdditionalData` will be useful in any case... including that update to the latest API. I'm still trying to figure out exact scenario this is trying to triage. Seems that whole problem started because we are (for some reason) doing GET/POST for updates... while we should be solving all this with simple PATCH/PUT.

My biggest problem with `AdditionalData` is that if you want to apply this pattern - you now need to have this collection on every update request... and you force every client implementation to cater to this. That's my primary motivation for joining this conversation - down the road when client implementations start - I'm sure they would rather not implement this.

Meeting is coming up tomorrow so I think it's best we discuss this there. I don't think we need too much structure... just enough to address scenarios that Nicolas sees as problematic, and I think quick live conversation will provide that.

**Comment by Kukks** [comment]
We aren't, we are using POST for create and PUT for update.

AdditionalData is completely optional on both the client and the server
though. The server collects any unidentified properties and adds them
there. And the client can just fill it in case the current client version
doesn't support any new server properties yet (I guess?). Point is no
client needs it even if its present in the c# client
